File: scripts/reorganize-price-list.py
#!/usr/bin/env python3
import json
import openpyxl
from openpyxl.utils import get_column_letter
from openpyxl.styles import Alignment, Font, PatternFill
from copy import copy

print('Reorganizing price list with full formatting preservation...')

# Read the JSON data
with open('public/price-data.json', 'r') as f:
    current_data = json.load(f)

# Load the original workbook with full formatting
original_wb = openpyxl.load_workbook('Archive/Bogner Price List 11.11.25.xlsx')
original_ws = original_wb.active

# Load again with data_only to get computed values for formulas
original_wb_values = openpyxl.load_workbook('Archive/Bogner Price List 11.11.25.xlsx', data_only=True)
original_ws_values = original_wb_values.active

# Create new workbook
new_wb = openpyxl.Workbook()
new_ws = new_wb.active
new_ws.title = 'Price List Reorganized'

# Helper function to find row by text in column C
def find_row_by_text(search_text, search_type='exact'):
    for row_idx, row in enumerate(original_ws.iter_rows(min_col=3, max_col=3), start=1):
        cell = row[0]
        if cell.value:
            cell_value = str(cell.value).strip()
            search_value = str(search_text).strip()

            if search_type == 'exact' and cell_value == search_value:
                return row_idx
            elif search_type == 'contains' and search_value in cell_value:
                return row_idx
    return -1

# Helper function to copy a row with all formatting
def copy_row(source_row_idx, target_row_idx):
    source_row = original_ws[source_row_idx]

    for col_idx, source_cell in enumerate(source_row, start=1):
        if col_idx > 3:  # Only copy columns A, B, C
            break

        target_cell = new_ws.cell(row=target_row_idx, column=col_idx)
        source_cell_value = original_ws_values.cell(row=source_row_idx, column=col_idx)

        # Copy value (use value workbook to avoid formulas)
        if source_cell_value.value is not None:
            target_cell.value = source_cell_value.value
        elif source_cell.value is not None and not (isinstance(source_cell.value, str) and source_cell.value.startswith('=')):
            target_cell.value = source_cell.value

        # Copy formatting carefully
        if source_cell.has_style:
            try:
                if source_cell.font:
                    target_cell.font = copy(source_cell.font)
            except:
                pass
            try:
                if source_cell.border:
                    target_cell.border = copy(source_cell.border)
            except:
                pass
            try:
                if source_cell.fill:
                    target_cell.fill = copy(source_cell.fill)
            except:
                pass
            try:
                if source_cell.number_format:
                    target_cell.number_format = source_cell.number_format
            except:
                pass
            try:
                if source_cell.alignment:
                    target_cell.alignment = copy(source_cell.alignment)
            except:
                pass

# Helper function to get all rows for a section
def get_section_rows(section_name):
    start_row = find_row_by_text(section_name, 'exact')
    if start_row == -1:
        return []

    rows = [start_row]

    # Continue until we hit another section header
    for row_idx in range(start_row + 1, original_ws.max_row + 1):
        cell_c = original_ws.cell(row=row_idx, column=3)
        cell_a = original_ws.cell(row=row_idx, column=1)

        # Stop if we hit a new section (no cost/unit and not a note)
        if cell_c.value and not cell_a.value:
            text = str(cell_c.value).strip()
            if not text.startswith('Note'):
                break

        rows.append(row_idx)

        # Stop if we've gone far enough and hit empty cells
        if not cell_c.value and not cell_a.value:
            next_cell = original_ws.cell(row=row_idx + 1, column=3)
            if not next_cell.value:
                break

    return rows

# Build reorganized data structure
reorganized_data = {
    'categories': []
}

# NEW CONSTRUCTION
new_construction = {'name': 'NEW CONSTRUCTION', 'sections': []}

# Phase headers and sections as defined in the JS script
sections_order = [
    {'type': 'phase', 'name': 'PHASE 1: PROJECT PLANNING & SITE PREPARATION'},
    {'type': 'section', 'name': 'Pool Base Cost'},
    {'type': 'section', 'name': 'Pool Depths'},
    {'type': 'section', 'name': 'Zone Charges'},
    {'type': 'section', 'name': 'Structural Engineering'},
    {'type': 'section', 'name': 'Excavation'},

    {'type': 'phase', 'name': 'PHASE 2: POOL SHELL CONSTRUCTION'},
    {'type': 'section', 'name': 'Shotcrete'},
    {'type': 'section', 'name': 'Raised Bond Beam'},
    {'type': 'section', 'name': 'Wing Walls'},
    {'type': 'section', 'name': 'Vanishing Edge'},
    {'type': 'section', 'name': 'Spa Base Cost'},
    {'type': 'section', 'name': 'Spa Extras'},
    {'type': 'section', 'name': 'Spa Walls'},
    {'type': 'section', 'name': 'Spa Only'},

    {'type': 'phase', 'name': 'PHASE 3: PLUMBING SYSTEMS'},
    {'type': 'section', 'name': 'Pool Plumbing'},
    {'type': 'section', 'name': 'Spa Plumbing'},
    {'type': 'section', 'name': 'In Floor Cleaner'},
    {'type': 'section', 'name': 'Pool Sweeps'},
    {'type': 'section', 'name': 'Drains'},
    {'type': 'section', 'name': 'Gas Lines'},

    {'type': 'phase', 'name': 'PHASE 4: EQUIPMENT INSTALLATION'},
    {'type': 'section', 'name': 'Heaters'},
    {'type': 'section', 'name': 'Sanitizers'},

    {'type': 'phase', 'name': 'PHASE 5: ELECTRICAL SYSTEMS'},
    {'type': 'section', 'name': 'Electrical'},
    {'type': 'section', 'name': 'Remote Controls'},

    {'type': 'phase', 'name': 'PHASE 6: INTERIOR FINISHES'},
    {'type': 'section', 'name': 'Tile'},
    {'type': 'section', 'name': 'Coping'},
    {'type': 'section', 'name': 'Pebble/Quartz/Colored Plaster'},

    {'type': 'phase', 'name': 'PHASE 7: POOL FEATURES & ACCESSORIES'},
    {'type': 'section', 'name': 'Pool Extras'},
    {'type': 'section', 'name': 'Water Features'},
    {'type': 'section', 'name': 'Motorized Pool Cover'},

    {'type': 'phase', 'name': 'PHASE 8: DECKING & HARDSCAPE'},
    {'type': 'section', 'name': 'Decking'},
    {'type': 'section', 'name': 'Footings'},
    {'type': 'section', 'name': 'Step Risers'},
    {'type': 'section', 'name': 'Pavers'},

    {'type': 'phase', 'name': 'PHASE 9: WALLS & STRUCTURES'},
    {'type': 'section', 'name': 'Walls'},
    {'type': 'section', 'name': 'Wall Caps'},
    {'type': 'section', 'name': 'Columns'},
    {'type': 'section', 'name': 'Column Caps'},
    {'type': 'section', 'name': 'Facing and Veneer Not Raised Bond Beam'},

    {'type': 'phase', 'name': 'PHASE 10: OUTDOOR LIVING FEATURES'},
    {'type': 'section', 'name': 'Bbqs'},
    {'type': 'section', 'name': 'Fire Pits and Bowls'},
    {'type': 'section', 'name': 'Real Rock'},
    {'type': 'section', 'name': 'Artificial Rock'},
    {'type': 'section', 'name': 'Solar'},
    {'type': 'section', 'name': 'Alumawood Patio Covers'},

    {'type': 'phase', 'name': 'PHASE 11: FENCING & SAFETY'},
    {'type': 'section', 'name': 'Fencing'},
]

# Copy column widths
for col_idx in range(1, 4):
    col_letter = get_column_letter(col_idx)
    if original_ws.column_dimensions[col_letter].width:
        new_ws.column_dimensions[col_letter].width = original_ws.column_dimensions[col_letter].width

# Make sure column C is wide enough for phase headers
if new_ws.column_dimensions['C'].width and new_ws.column_dimensions['C'].width < 60:
    new_ws.column_dimensions['C'].width = 60

# Start copying rows
current_target_row = 1

# Copy header row (Cost, Factor, Item)
header_row = 1
for row in original_ws.iter_rows(min_row=1, max_row=10):
    if row[0].value == 'Cost' or row[2].value == 'Item':
        header_row = row[0].row
        break

copy_row(header_row, current_target_row)
current_target_row += 1

# Copy "Last Updated" row
last_updated_row = find_row_by_text('Last Updated', 'contains')
if last_updated_row > 0:
    copy_row(last_updated_row, current_target_row)
    current_target_row += 1

# Copy NEW CONSTRUCTION category row
category_row = find_row_by_text('New Construction', 'exact')
print(f'Found New Construction at row: {category_row}')
if category_row > 0:
    copy_row(category_row, current_target_row)
    # Change text to uppercase
    new_ws.cell(row=current_target_row, column=3).value = 'NEW CONSTRUCTION'
    current_target_row += 1
else:
    print('ERROR: Could not find New Construction category!')

# Process all NEW CONSTRUCTION sections
for item in sections_order:
    if item['type'] == 'phase':
        # Copy category row format for phase header
        phase_template_row = find_row_by_text('New Construction', 'exact')
        if phase_template_row <= 0:
            phase_template_row = find_row_by_text('Remodel', 'exact')

        if phase_template_row > 0:
            copy_row(phase_template_row, current_target_row)
            # Clear columns A and B, update column C with phase name
            phase_cell_a = new_ws.cell(row=current_target_row, column=1)
            phase_cell_b = new_ws.cell(row=current_target_row, column=2)
            phase_cell_c = new_ws.cell(row=current_target_row, column=3)

            phase_cell_a.value = None
            phase_cell_b.value = None
            phase_cell_c.value = item['name']

            # Customize phase header formatting using website color scheme (slate-700)
            phase_cell_c.font = Font(name='Arial', size=14, bold=True, color='FFFFFF')
            phase_cell_c.fill = PatternFill(start_color='334155', end_color='334155', fill_type='solid')  # slate-700
            phase_cell_c.alignment = Alignment(horizontal='center', vertical='center')

            print(f'Created phase header: {item["name"]}')
            current_target_row += 1
    else:
        # Copy entire section with all its rows
        section_rows = get_section_rows(item['name'])
        if section_rows:
            print(f'Copying section "{item["name"]}" - {len(section_rows)} rows')
            for source_row in section_rows:
                copy_row(source_row, current_target_row)
                current_target_row += 1
        else:
            print(f'WARNING: Section "{item["name"]}" not found!')

# Copy REMODEL category row
remodel_row = find_row_by_text('Remodel', 'exact')
print(f'Found Remodel at row: {remodel_row}')
if remodel_row > 0:
    copy_row(remodel_row, current_target_row)
    # Change text to uppercase
    new_ws.cell(row=current_target_row, column=3).value = 'REMODEL'
    current_target_row += 1

# REMODEL sections (simplified - copy all remodel sections as they are)
remodel_sections = [
    {'type': 'phase', 'name': 'REMODEL PHASE 1: DEMOLITION & PREPARATION'},
    {'type': 'section', 'name': 'Strip Plaster'},
    {'type': 'section', 'name': 'Dump Fees'},
    {'type': 'phase', 'name': 'REMODEL PHASE 2: SURFACE PREPARATION'},
    {'type': 'section', 'name': 'Cut Tile if Tile Stays on Pool and Spa'},
    {'type': 'section', 'name': 'Tile up to Group 5'},
    {'type': 'section', 'name': 'Coping'},
    {'type': 'phase', 'name': 'REMODEL PHASE 3: NEW FINISHES'},
    {'type': 'section', 'name': 'White Plaster'},
    {'type': 'section', 'name': 'Pebble Finish'},
    {'type': 'section', 'name': 'Sparkle Quartz'},
    {'type': 'phase', 'name': 'REMODEL PHASE 4: SYSTEMS UPGRADES'},
    {'type': 'section', 'name': 'Plumbing'},
    {'type': 'section', 'name': 'Electrical'},
    {'type': 'section', 'name': 'Equipment'},
    {'type': 'phase', 'name': 'REMODEL PHASE 5: DECKING'},
    {'type': 'section', 'name': 'Decking'},
]

for item in remodel_sections:
    if item['type'] == 'phase':
        # Copy category row format for phase header
        phase_row = find_row_by_text('Remodel', 'exact')
        if phase_row > 0:
            copy_row(phase_row, current_target_row)
            # Clear columns A and B, update column C with phase name
            phase_cell_a = new_ws.cell(row=current_target_row, column=1)
            phase_cell_b = new_ws.cell(row=current_target_row, column=2)
            phase_cell_c = new_ws.cell(row=current_target_row, column=3)

            phase_cell_a.value = None
            phase_cell_b.value = None
            phase_cell_c.value = item['name']

            # Customize remodel phase header formatting using website color scheme (slate-600 - lighter grey)
            phase_cell_c.font = Font(name='Arial', size=14, bold=True, color='FFFFFF')
            phase_cell_c.fill = PatternFill(start_color='475569', end_color='475569', fill_type='solid')  # slate-600
            phase_cell_c.alignment = Alignment(horizontal='center', vertical='center')

            print(f'Created remodel phase header: {item["name"]}')
            current_target_row += 1
    else:
        # Copy entire section with all its rows
        section_rows = get_section_rows(item['name'])
        if section_rows:
            print(f'Copying remodel section "{item["name"]}" - {len(section_rows)} rows')
            for source_row in section_rows:
                copy_row(source_row, current_target_row)
                current_target_row += 1
        else:
            print(f'WARNING: Remodel section "{item["name"]}" not found!')

# Save the new workbook
new_wb.save('Archive/Bogner Price List - Reorganized.xlsx')

print('✓ Reorganization complete with full formatting preserved!')
print(f'✓ File saved to: Archive/Bogner Price List - Reorganized.xlsx')
print(f'✓ Total rows copied: {current_target_row - 1}')

File: scripts/reorganize-price-list.js
const XLSX = require('xlsx');
const fs = require('fs');
const path = require('path');

console.log('Reorganizing price list by copying and rearranging original Excel rows...');

// Read the original Excel file with full formatting preservation
const originalExcelPath = path.join(__dirname, '../Archive/Bogner Price List 11.11.25.xlsx');
const originalWorkbook = XLSX.readFile(originalExcelPath, {
  cellStyles: true,
  cellHTML: false,
  cellFormula: true,
  cellDates: true
});
const originalSheet = originalWorkbook.Sheets[originalWorkbook.SheetNames[0]];

// Read the current JSON data
const jsonPath = path.join(__dirname, '../public/price-data.json');
const currentData = JSON.parse(fs.readFileSync(jsonPath, 'utf8'));

// Create new reorganized structure
const reorganizedData = {
  title: 'Bogner Price List - Reorganized by Construction Phase',
  lastUpdated: currentData.lastUpdated,
  categories: []
};

// Helper function to find section by name in current data
function findSection(sectionName) {
  for (const category of currentData.categories) {
    const section = category.sections.find(s => s.name === sectionName);
    if (section) return JSON.parse(JSON.stringify(section)); // Deep copy
  }
  return null;
}

// Helper function to find multiple sections
function findSections(...sectionNames) {
  return sectionNames.map(name => findSection(name)).filter(s => s !== null);
}

// NEW CONSTRUCTION CATEGORY
const newConstruction = {
  name: 'NEW CONSTRUCTION',
  sections: []
};

// PHASE 1: Project Planning & Site Preparation
newConstruction.sections.push({
  name: '=== PHASE 1: PROJECT PLANNING & SITE PREPARATION ===',
  items: [],
  notes: ['This phase includes initial project setup, engineering, and site access preparation']
});

// Pool Base Pricing & Specifications
const poolBase = findSection('Pool Base Cost');
if (poolBase) {
  poolBase.name = 'Pool Base Pricing & Specifications';
  newConstruction.sections.push(poolBase);
}

const poolDepths = findSection('Pool Depths');
if (poolDepths) newConstruction.sections.push(poolDepths);

const zoneCharges = findSection('Zone Charges');
if (zoneCharges) newConstruction.sections.push(zoneCharges);

const structuralEng = findSection('Structural Engineering');
if (structuralEng) newConstruction.sections.push(structuralEng);

// Excavation & Site Work
const excavation = findSection('Excavation');
if (excavation) {
  excavation.name = 'Excavation & Site Work';
  newConstruction.sections.push(excavation);
}

// PHASE 2: Pool Shell Construction
newConstruction.sections.push({
  name: '=== PHASE 2: POOL SHELL CONSTRUCTION ===',
  items: [],
  notes: ['Shotcrete, structural elements, and shell finishing']
});

const shotcrete = findSection('Shotcrete');
if (shotcrete) {
  shotcrete.name = 'Pool Structure - Shotcrete & Steel';
  newConstruction.sections.push(shotcrete);
}

const raisedBondBeam = findSection('Raised Bond Beam');
if (raisedBondBeam) newConstruction.sections.push(raisedBondBeam);

const wingWalls = findSection('Wing Walls');
if (wingWalls) newConstruction.sections.push(wingWalls);

const vanishingEdge = findSection('Vanishing Edge');
if (vanishingEdge) newConstruction.sections.push(vanishingEdge);

// Spa Construction (consolidated)
newConstruction.sections.push({
  name: 'Spa Construction (All Elements)',
  items: [],
  notes: ['Consolidated spa base, extras, and walls for easier estimation']
});

const spaBase = findSection('Spa Base Cost');
if (spaBase) newConstruction.sections.push(spaBase);

const spaExtras = findSection('Spa Extras');
if (spaExtras) newConstruction.sections.push(spaExtras);

const spaWalls = findSection('Spa Walls');
if (spaWalls) newConstruction.sections.push(spaWalls);

const spaOnly = findSection('Spa Only');
if (spaOnly) newConstruction.sections.push(spaOnly);

// PHASE 3: Plumbing Systems
newConstruction.sections.push({
  name: '=== PHASE 3: PLUMBING SYSTEMS ===',
  items: [],
  notes: ['All water circulation, drainage, and specialty plumbing']
});

const poolPlumbing = findSection('Pool Plumbing');
if (poolPlumbing) newConstruction.sections.push(poolPlumbing);

const spaPlumbing = findSection('Spa Plumbing');
if (spaPlumbing) newConstruction.sections.push(spaPlumbing);

const inFloorCleaner = findSection('In Floor Cleaner');
if (inFloorCleaner) {
  inFloorCleaner.name = 'Specialty Plumbing - In Floor Cleaning';
  newConstruction.sections.push(inFloorCleaner);
}

const poolSweeps = findSection('Pool Sweeps');
if (poolSweeps) {
  poolSweeps.name = 'Specialty Plumbing - Pool Sweeps';
  newConstruction.sections.push(poolSweeps);
}

const drains = findSection('Drains');
if (drains) {
  drains.name = 'Specialty Plumbing - Drainage Systems';
  newConstruction.sections.push(drains);
}

const gasLines = findSection('Gas Lines');
if (gasLines) {
  gasLines.name = 'Gas Line Systems';
  newConstruction.sections.push(gasLines);
}

// PHASE 4: Equipment Installation
newConstruction.sections.push({
  name: '=== PHASE 4: EQUIPMENT INSTALLATION ===',
  items: [],
  notes: ['Filtration, heating, and water treatment systems']
});

const heaters = findSection('Heaters');
if (heaters) {
  heaters.name = 'Heating Systems';
  newConstruction.sections.push(heaters);
}

const sanitizers = findSection('Sanitizers');
if (sanitizers) {
  sanitizers.name = 'Water Treatment & Sanitization';
  newConstruction.sections.push(sanitizers);
}

// PHASE 5: Electrical Systems
newConstruction.sections.push({
  name: '=== PHASE 5: ELECTRICAL SYSTEMS ===',
  items: [],
  notes: ['All electrical runs, lights, and controls']
});

const electrical = findSection('Electrical');
if (electrical) {
  electrical.name = 'Pool & Feature Electrical';
  newConstruction.sections.push(electrical);
}

const remoteControls = findSection('Remote Controls');
if (remoteControls) newConstruction.sections.push(remoteControls);

// PHASE 6: Interior Finishes
newConstruction.sections.push({
  name: '=== PHASE 6: INTERIOR FINISHES ===',
  items: [],
  notes: ['Tile, coping, and plaster/surface finishes']
});

const tile = findSection('Tile');
if (tile) {
  tile.name = 'Tile Work';
  newConstruction.sections.push(tile);
}

const coping = findSection('Coping');
if (coping) newConstruction.sections.push(coping);

const plaster = findSection('Pebble/Quartz/Colored Plaster');
if (plaster) {
  plaster.name = 'Plaster & Surface Finishes';
  newConstruction.sections.push(plaster);
}

// PHASE 7: Pool Features & Accessories
newConstruction.sections.push({
  name: '=== PHASE 7: POOL FEATURES & ACCESSORIES ===',
  items: [],
  notes: ['Water features, safety features, and pool accessories']
});

const poolExtras = findSection('Pool Extras');
if (poolExtras) {
  poolExtras.name = 'Pool Entry & Safety Features';
  newConstruction.sections.push(poolExtras);
}

const waterFeatures = findSection('Water Features');
if (waterFeatures) newConstruction.sections.push(waterFeatures);

const motorizedCover = findSection('Motorized Pool Cover');
if (motorizedCover) newConstruction.sections.push(motorizedCover);

// PHASE 8: Decking & Hardscape
newConstruction.sections.push({
  name: '=== PHASE 8: DECKING & HARDSCAPE ===',
  items: [],
  notes: ['Concrete decking, footings, steps, and pavers']
});

const decking = findSection('Decking');
if (decking) {
  decking.name = 'Concrete Decking';
  newConstruction.sections.push(decking);
}

const footings = findSection('Footings');
if (footings) newConstruction.sections.push(footings);

const stepRisers = findSection('Step Risers');
if (stepRisers) {
  stepRisers.name = 'Steps & Risers';
  newConstruction.sections.push(stepRisers);
}

const pavers = findSection('Pavers');
if (pavers) newConstruction.sections.push(pavers);

// PHASE 9: Walls & Structures
newConstruction.sections.push({
  name: '=== PHASE 9: WALLS & STRUCTURES ===',
  items: [],
  notes: ['Walls, columns, caps, and veneer work']
});

const walls = findSection('Walls');
if (walls) {
  walls.name = 'Freestanding & Retaining Walls';
  newConstruction.sections.push(walls);
}

const wallCaps = findSection('Wall Caps');
if (wallCaps) newConstruction.sections.push(wallCaps);

const columns = findSection('Columns');
if (columns) newConstruction.sections.push(columns);

const columnCaps = findSection('Column Caps');
if (columnCaps) newConstruction.sections.push(columnCaps);

const facing = findSection('Facing and Veneer Not Raised Bond Beam');
if (facing) {
  facing.name = 'Facing & Veneer';
  newConstruction.sections.push(facing);
}

// PHASE 10: Outdoor Living Features
newConstruction.sections.push({
  name: '=== PHASE 10: OUTDOOR LIVING FEATURES ===',
  items: [],
  notes: ['BBQs, fire features, and specialty outdoor amenities']
});

const bbqs = findSection('Bbqs');
if (bbqs) {
  bbqs.name = 'BBQ Islands';
  newConstruction.sections.push(bbqs);
}

const firePits = findSection('Fire Pits and Bowls');
if (firePits) newConstruction.sections.push(firePits);

const realRock = findSection('Real Rock');
if (realRock) newConstruction.sections.push(realRock);

const artificialRock = findSection('Artificial Rock');
if (artificialRock) newConstruction.sections.push(artificialRock);

const solar = findSection('Solar');
if (solar) newConstruction.sections.push(solar);

const alumawood = findSection('Alumawood Patio Covers');
if (alumawood) newConstruction.sections.push(alumawood);

// PHASE 11: Fencing & Safety
newConstruction.sections.push({
  name: '=== PHASE 11: FENCING & SAFETY ===',
  items: [],
  notes: ['Fence installation and safety compliance']
});

const fencing = findSection('Fencing');
if (fencing) {
  fencing.name = 'Fencing & Gates';
  newConstruction.sections.push(fencing);
}

// Add new construction category
reorganizedData.categories.push(newConstruction);

// REMODEL CATEGORY
const remodel = {
  name: 'REMODEL',
  sections: []
};

// REMODEL Phase 1: Demolition & Preparation
remodel.sections.push({
  name: '=== REMODEL PHASE 1: DEMOLITION & PREPARATION ===',
  items: [],
  notes: ['Strip and removal of existing surfaces']
});

const stripPlaster = findSection('Strip Plaster');
if (stripPlaster) {
  stripPlaster.name = 'Strip & Removal';
  remodel.sections.push(stripPlaster);
}

const dumpFees = findSection('Dump Fees');
if (dumpFees) remodel.sections.push(dumpFees);

// REMODEL Phase 2: Surface Preparation
remodel.sections.push({
  name: '=== REMODEL PHASE 2: SURFACE PREPARATION ===',
  items: [],
  notes: ['Tile cutting and coping preparation']
});

const cutTile = findSection('Cut Tile if Tile Stays on Pool and Spa');
if (cutTile) {
  cutTile.name = 'Tile Work (if tile stays)';
  remodel.sections.push(cutTile);
}

const remodelTile = findSection('Tile up to Group 5');
if (remodelTile) {
  remodelTile.name = 'Tile Replacement';
  remodel.sections.push(remodelTile);
}

const remodelCoping = findSection('Coping');
if (remodelCoping) {
  remodelCoping.name = 'Coping Replacement';
  remodel.sections.push(remodelCoping);
}

// REMODEL Phase 3: New Finishes
remodel.sections.push({
  name: '=== REMODEL PHASE 3: NEW FINISHES ===',
  items: [],
  notes: ['Plaster and surface finish application']
});

const whitePlaster = findSection('White Plaster');
if (whitePlaster) {
  whitePlaster.name = 'Plaster & Surface Finishes';
  remodel.sections.push(whitePlaster);
}

const pebbleFinish = findSection('Pebble Finish');
if (pebbleFinish) remodel.sections.push(pebbleFinish);

const sparkleQuartz = findSection('Sparkle Quartz');
if (sparkleQuartz) remodel.sections.push(sparkleQuartz);

// REMODEL Phase 4: Systems Upgrades
remodel.sections.push({
  name: '=== REMODEL PHASE 4: SYSTEMS UPGRADES ===',
  items: [],
  notes: ['Plumbing, electrical, and equipment updates']
});

const remodelPlumbing = findSection('Plumbing');
if (remodelPlumbing) {
  remodelPlumbing.name = 'Plumbing Upgrades';
  remodel.sections.push(remodelPlumbing);
}

const remodelElectrical = findSection('Electrical');
if (remodelElectrical) {
  remodelElectrical.name = 'Electrical Upgrades';
  remodel.sections.push(remodelElectrical);
}

const remodelEquipment = findSection('Equipment');
if (remodelEquipment) {
  remodelEquipment.name = 'Equipment Replacement';
  remodel.sections.push(remodelEquipment);
}

// REMODEL Phase 5: Decking
remodel.sections.push({
  name: '=== REMODEL PHASE 5: DECKING ===',
  items: [],
  notes: ['Use new construction pricing + $1 PSF']
});

const remodelDecking = findSection('Decking');
if (remodelDecking) {
  remodelDecking.name = 'Decking Work';
  remodel.sections.push(remodelDecking);
}

// Add remodel category
reorganizedData.categories.push(remodel);

// Helper function to find row index by text in column C
function findRowByText(searchText, searchType = 'exact') {
  const range = XLSX.utils.decode_range(originalSheet['!ref']);
  for (let R = range.s.r; R <= range.e.r; R++) {
    const cellAddress = XLSX.utils.encode_cell({ r: R, c: 2 }); // Column C
    const cell = originalSheet[cellAddress];
    if (cell && cell.v) {
      const cellValue = String(cell.v).trim();
      const searchValue = String(searchText).trim();

      if (searchType === 'exact' && cellValue === searchValue) {
        return R;
      } else if (searchType === 'contains' && cellValue.includes(searchValue)) {
        return R;
      }
    }
  }
  return -1;
}

// Helper function to copy entire row with all formatting
function copyRow(sourceRow, targetRow) {
  const range = XLSX.utils.decode_range(originalSheet['!ref']);
  const newCells = {};

  for (let C = range.s.c; C <= range.e.c; C++) {
    const sourceAddr = XLSX.utils.encode_cell({ r: sourceRow, c: C });
    const targetAddr = XLSX.utils.encode_cell({ r: targetRow, c: C });

    if (originalSheet[sourceAddr]) {
      // Deep copy the cell with all properties (value, style, format, etc.)
      newCells[targetAddr] = JSON.parse(JSON.stringify(originalSheet[sourceAddr]));
    }
  }

  return newCells;
}

// Helper function to get all rows for a section (section name + notes + items)
function getSectionRows(sectionName) {
  const startRow = findRowByText(sectionName, 'exact');
  if (startRow === -1) return [];

  const rows = [startRow];
  const range = XLSX.utils.decode_range(originalSheet['!ref']);

  // Continue until we hit another section header or category
  for (let R = startRow + 1; R <= range.e.r; R++) {
    const cellC = originalSheet[XLSX.utils.encode_cell({ r: R, c: 2 })];
    const cellA = originalSheet[XLSX.utils.encode_cell({ r: R, c: 0 })];

    // Stop if we hit a new section (no cost/unit and not a note)
    if (cellC && cellC.v && !cellA) {
      const text = String(cellC.v).trim();
      if (!text.startsWith('Note')) {
        break;
      }
    }

    rows.push(R);

    // Stop if we've gone far enough and hit empty cells
    if (!cellC && !cellA) {
      const nextCell = originalSheet[XLSX.utils.encode_cell({ r: R + 1, c: 2 })];
      if (!nextCell) break;
    }
  }

  return rows;
}

// Create new worksheet by copying and rearranging original rows
const newSheet = {};
let currentTargetRow = 0;

// Copy header row (row 0)
Object.assign(newSheet, copyRow(0, currentTargetRow++));

// Copy "Last Updated" row (row 1)
const lastUpdatedRow = findRowByText('Last Updated', 'contains');
if (lastUpdatedRow >= 0) {
  Object.assign(newSheet, copyRow(lastUpdatedRow, currentTargetRow++));
}

// Process reorganized data
reorganizedData.categories.forEach(category => {
  // Find and copy category row
  const categoryRow = findRowByText(category.name, 'exact');
  if (categoryRow >= 0) {
    Object.assign(newSheet, copyRow(categoryRow, currentTargetRow++));
  }

  category.sections.forEach(section => {
    // Handle phase headers (create new formatted row)
    if (section.name.includes('===')) {
      // Copy a category row format for phase header
      const phaseHeaderRow = findRowByText('NEW CONSTRUCTION', 'exact');
      if (phaseHeaderRow >= 0) {
        const phaseCells = copyRow(phaseHeaderRow, currentTargetRow);
        // Update the text in column C
        const cellC = XLSX.utils.encode_cell({ r: currentTargetRow, c: 2 });
        if (phaseCells[cellC]) {
          phaseCells[cellC].v = section.name;
          phaseCells[cellC].w = section.name;
        }
        Object.assign(newSheet, phaseCells);
        currentTargetRow++;
      }
    } else {
      // Find and copy all rows for this section
      const sectionRows = getSectionRows(section.name);
      sectionRows.forEach(sourceRow => {
        Object.assign(newSheet, copyRow(sourceRow, currentTargetRow++));
      });
    }
  });
});

// Set the range for the new sheet
newSheet['!ref'] = XLSX.utils.encode_range({
  s: { r: 0, c: 0 },
  e: { r: currentTargetRow - 1, c: 2 }
});

// Copy column widths
if (originalSheet['!cols']) {
  newSheet['!cols'] = JSON.parse(JSON.stringify(originalSheet['!cols']));
}

// Copy other sheet properties
if (originalSheet['!rows']) {
  newSheet['!rows'] = JSON.parse(JSON.stringify(originalSheet['!rows']));
}
if (originalSheet['!merges']) {
  newSheet['!merges'] = JSON.parse(JSON.stringify(originalSheet['!merges']));
}

// Create workbook and add the reorganized sheet
const workbook = XLSX.utils.book_new();
XLSX.utils.book_append_sheet(workbook, newSheet, 'Price List Reorganized');

// Write to file
const outputPath = path.join(__dirname, '../Archive/Bogner Price List - Reorganized.xlsx');
XLSX.writeFile(workbook, outputPath, {
  cellStyles: true,
  bookSST: true,
  bookType: 'xlsx'
});

console.log('✓ Reorganization complete!');
console.log(`✓ File saved to: ${outputPath}`);
console.log(`✓ Categories: ${reorganizedData.categories.length}`);
console.log(`✓ New Construction Sections: ${reorganizedData.categories[0].sections.length}`);
console.log(`✓ Remodel Sections: ${reorganizedData.categories[1].sections.length}`);
